### PR TITLE
feat: add compactor concurrency within single partition's plans

### PR DESCRIPTION
**Pull request by Joe-Blount**
This adds concurrency within a single partition's DF plan execution.  The existing job semaphore still applies.  This PR does not allow higher maximum concurrency to DF (which is a good thing - we have one reliable knot to turn to control concurrency in DF plan execution).  This PR allows a backlogged partition to make more use of the DF concurrency limit.

The consequence of adding this concurrency is expected to be:
- when 1 (or few) partitions are highly backlogged, and many other partitions are caught up, this will allow the highly backlogged partition to get a higher share of the job semaphore limit, helping it catch up more quickly (at the expense of the partitions that are mostly caught up, who's compactions aren't as high of ROI).
- When 1 (or few) partitions are highly backlogged, and everything else needs no compaction, this will allow a single partition to make use of the allowed concurrency, rather than the compactor requiring multiple partitions before it runs efficiently.

Helps: https://github.com/influxdata/EAR/issues/4168
Helps: https://github.com/influxdata/idpe/issues/17480

The insta test churn is reordering of compaction plan's completions, because now they get to run concurrently.  

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: compactor2/src/driver.rs [hunk 1]
```diff
@@ -327,25 +327,25 @@ async fn run_plans(
...
-    {
-        created_file_params.extend(
-            execute_plan(
-                plan_ir,
-                partition_info,
-                components,
-                Arc::clone(&job_semaphore),
-            )
-            .await?,
+    let created_file_params: Vec<Vec<_>> = stream::iter(
```
Comment (NGA-TRAN) [inline compactor2/src/driver.rs:331]: You may want to add (temporarily) log or something like that for you to know how many plans here. If IIRC, the number of plans is the number of split plans or the number of output files (each file is a split). If this is not a actually vertical split, I do no think we will have many plans for the concurrency to help much here

File: compactor2/tests/layouts/backfill.rs [hunk 2]
```diff
@@ -759,76 +759,76 @@ async fn random_backfill_empty_partition() {
...
-    - "L0.?[659,670] 1.04us 143.55kb                                                                                      |L0.?|"
-    - "**** Simulation run 72, type=split(ReduceOverlap)(split_times=[329]). 1 Input Files, 3.33mb total:"
-    - "L0, all files 3.33mb                                                                                               "
-    - "L0.165[42,356] 1.04us    |-----------------------------------------L0.165-----------------------------------------|"
-    - "**** 2 Output Files (parquet_file_id not yet assigned), 3.33mb total:"
-    - "L0                                                                                                                 "
-    - "L0.?[42,329] 1.04us 3.04mb|--------------------------------------L0.?--------------------------------------|        "
-    - "L0.?[330,356] 1.04us 292.88kb                                                                                  |L0.?-| "
-    - "**** Simulation run 73, type=split(ReduceOverlap)(split_times=[658]). 1 Input Files, 3.32mb total:"
+    - "**** Simulation run 71, type=split(ReduceOverlap)(split_times=[658]). 1 Input Files, 3.32mb total:"
```
Comment (NGA-TRAN) [inline compactor2/tests/layouts/backfill.rs:762]: Are the changes in the tests due to different splitting order due to concurency? I do not look into the tests but I think you may want to verify that they are the same and just changes in order of compaction

**Review by NGA-TRAN — APPROVED** [review]
The code looks good to me.  Log/metrics will help to see how often this helps

**Comment by NGA-TRAN** [comment]
@Joe-Blount One more place you can add concurrency is inside this loop. The branches can be run in parallel

File: compactor2/tests/layouts/backfill.rs (hunk 2 above)
Comment (Joe-Blount) [reply compactor2/tests/layouts/backfill.rs:762]: I think its reordering.  Some of the plans complete in different orders, so some of the generated file names change.  Note the the bytes written didn't change in the final output.

**Pull request merged**